File: projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/gpu/jpa/NodeWithGPUs.java
/*
 * Copyright 2023-2024 Broadcom
 * SPDX-License-Identifier: Apache-2.0
 */

package com.vmware.taurus.gpu.jpa;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import javax.persistence.*;
import java.util.HashSet;
import java.util.Set;


@Data
@AllArgsConstructor
@NoArgsConstructor
@Entity
public class NodeWithGPUs {

    @Id
    private String name;

    private int deviceCount;

    @OneToMany(mappedBy = "nodeWithGPUs", cascade = CascadeType.ALL, fetch = FetchType.EAGER)
    private Set<GpuConsumingJob> gpuConsumingJobs = new HashSet<>();


    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (!(obj instanceof NodeWithGPUs other)) {
            return false;
        }
        return java.util.Objects.equals(name, other.name);
    }

    @Override
    public int hashCode() {
        return java.util.Objects.hash(name);
    }

}

File: projects/control-service/projects/gpu_management/src/main/kotlin/com/vmware/gpu/management/jpa/NodeWithGPUs.kt
package com.vmware.gpu.management.jpa

import com.vmware.gpu.management.api.DeviceRequirements
import com.vmware.gpu.management.api.DeviceType
import jakarta.persistence.CascadeType
import jakarta.persistence.Entity
import jakarta.persistence.FetchType
import jakarta.persistence.Id
import jakarta.persistence.OneToMany
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository
import java.util.*




@Entity
class NodeWithGPUs(@Id val name: String,
                   val deviceCount: Int,
                   val deviceType: DeviceType) {
    @OneToMany(mappedBy = "nodeWithGPUs", cascade = [CascadeType.ALL], fetch = FetchType.EAGER)
    val gpuConsumingJobs: Set<GpuConsumingJob> = HashSet()
}
@Repository
interface NodeWithGPURepo : JpaRepository<NodeWithGPUs, String> {
    @Query(
        "SELECT n FROM NodeWithGPUs n LEFT JOIN n.gpuConsumingJobs g " +
                "WHERE n.deviceType = :deviceType " +
                "GROUP BY n HAVING n.deviceCount >=  :additionalResources+COALESCE(SUM(g.consumedResources), 0) " +
                "ORDER BY n.deviceCount - :additionalResources+COALESCE(SUM(g.consumedResources), 0) "
    )
    fun findNodesWithAvailableGPUs(additionalResources: Float,
                                   deviceType: DeviceType): List<NodeWithGPUs>
}


fun NodeWithGPURepo.findNodesWithAvailableGPUs(deviceRequirements: DeviceRequirements): List<NodeWithGPUs> = this.findNodesWithAvailableGPUs(
    deviceRequirements.amount ,  deviceRequirements.deviceType)
